src/runtime/refiner.ts

```diff
@@ -956,3 +956,138 @@ export class SQLExtracter {
     return `SELECT * FROM ${table}` + (filterTerms.length ? ` WHERE ${filterTerms.join(' AND ')}` : '') + ';';
   }
 }
+
+export class Fraction {
+  num: Polynomial;
+  den: Polynomial;
+
+  constructor(n?: Polynomial, d?: Polynomial) {
+    this.num = n ? Polynomial.copyOf(n) : new Polynomial();
+    this.den = d ? Polynomial.copyOf(d) : new Polynomial([1]);
+    if (this.den.isZero()) {
+      throw new Error('Division by zero.');
+    }
+    this.reduce();
+  }
+
+  static add(a: Fraction, b: Fraction): Fraction {
+    const den = Polynomial.multiply(a.den, b.den);
+    const num = Polynomial.add(Polynomial.multiply(a.num, b.den), Polynomial.multiply(b.num, a.den));
+    return new Fraction(num, den);
+  }
+
+  static negate(a: Fraction): Fraction {
+    return new Fraction(Polynomial.negate(a.num), a.den);
+  }
+
+  static subtract(a: Fraction, b: Fraction): Fraction {
+    const negB = Fraction.negate(b);
+    return Fraction.add(a, negB);
+  }
+
+  static multiply(a: Fraction, b: Fraction): Fraction {
+    return new Fraction(Polynomial.multiply(a.num, b.num), Polynomial.multiply(a.den, b.den));
+  }
+
+  static divide(a: Fraction, b: Fraction): Fraction {
+    const invB = new Fraction(b.den, b.num);
+    return Fraction.multiply(a, invB);
+  }
+
+  reduce() {
+    if (this.num.isZero()) {
+      this.den = new Polynomial([1]);
+      return;
+    }
+    if (this.num.isConstant() && this.den.isConstant()) {
+      this.num = new Polynomial([this.num.coeffs[0]/this.den.coeffs[0]]);
+      this.den = new Polynomial([1]);
+      return;
+    }
+    // TODO(ragdev): Fractions can be reduced further by factoring out the gcd of
+    // the coeffs in num and den, and then dividing the two. However, since the numbers are floating
+    // points, the precision and computation cost of gcd function will be a trade-off to consider.
+  }
+}
+
+export class Polynomial {
+  _coeffs: number[];
+
+  constructor(coeffs?: number[]) {
```

Comment:
I think this could be
```
constructor(coeffs?: number[] = [0]) {
   this.coeffs = coeffs;
}
```

Reply:
Fortunately javascript does the 'correct thing' here (in my view)

Unlike python
https://docs.python-guide.org/writing/gotchas/

Reply:
Ah yep, that is neater. 
This works:
```
constructor(coeffs: number[] = [0]) {
   this.coeffs = coeffs;
}
```
It doesn't let the parameter have a question mark and also an initializer.

Reply:
Whoops, my mistake, the question mark is meaningless with the default value.